### PR TITLE
Processor changes

### PR DESCRIPTION
**Pull request by Nykakin**
We'd like to remove processors from our utils and replace them with `zyte-common-items` counterparts. There are some functionality missing, however.

Summary of changes:
* Add new processor, `images_processor`, that attempts to convert input to list of `zyte_common_items.Image` instances
* Make `brand_processor` return `zyte_common_items.Brand` instances instead of strings
* Allow `brand_processor` to take string as an input argument
* Allow `price_processor` to take ~~string or~~ number as input argument (resolves https://github.com/zytedata/zyte-common-items/issues/94)

### CODE REVIEW

File: tests/test_pages_price.py [hunk 1]
```diff
@@ -62,21 +62,21 @@ def price(self):
...
-    assert page.price == "$13.2"
+    assert page.price == "13.20"
     assert page.call_count == 1
     assert page.currency is None
-    assert await page.currencyRaw is None
-    assert page.call_count == 2  # we want this to be 1
-    assert await page.currencyRaw is None
-    assert page.call_count == 2  # we want this to be 1
+    assert await page.currencyRaw == "$"
+    assert page.call_count == 1  # we want this to be 1
```
Comment (wRAR) [inline tests/test_pages_price.py:69]: I wonder why did the count change and what should be the new expectation in the comment.

File: zyte_common_items/processors.py [hunk 2]
```diff
@@ -104,50 +107,80 @@ def _from_zp_breadcrumb(value: zp_Breadcrumb) -> Breadcrumb:
     return results
 
 
-@only_handle_nodes
-def brand_processor(value: Union[Selector, HtmlElement], page: Any) -> Any:
+def brand_processor(value: Any, page: Any) -> Union[Brand, None]:
```
[the remark this answers is not in the document]
Comment (Nykakin) [reply zyte_common_items/processors.py]: Changed in ba1a5e93749c65eac05e4879588f87695aca818e

File: zyte_common_items/processors.py [hunk 3]
```diff
@@ -104,50 +107,80 @@ def _from_zp_breadcrumb(value: zp_Breadcrumb) -> Breadcrumb:
...
-    Supported inputs are :class:`~parsel.selector.Selector`,
-    :class:`~parsel.selector.SelectorList` and :class:`~lxml.html.HtmlElement`.
-    Other inputs are returned as is.
+    If inputs are either :class:`~parsel.selector.Selector`,
+    :class:`~parsel.selector.SelectorList` or :class:`~lxml.html.HtmlElement`, attempts
+    to extract brand data from it.
+
+    If value is a string, use it to create brand object instance
+
+    Other inputs are returned unchanged
```
[the remark this answers is not in the document]
Comment (Nykakin) [reply zyte_common_items/processors.py]: Changed in Changed in ba1a5e93749c65eac05e4879588f87695aca818e

File: tests/test_pages_price.py (hunk 1 above)
Comment (Nykakin) [reply tests/test_pages_price.py:69]: Yes, I've noticed that comment info doesn't match test values in places in master. Not sure if this is a proper behavior.

On master such difference can be found in 5 places:
* https://github.com/zytedata/zyte-common-items/blob/main/tests/test_pages_price.py#L41 
* https://github.com/zytedata/zyte-common-items/blob/main/tests/test_pages_price.py#L69
* https://github.com/zytedata/zyte-common-items/blob/main/tests/test_pages_price.py#L71
* https://github.com/zytedata/zyte-common-items/blob/main/tests/test_pages_price.py#L80
* https://github.com/zytedata/zyte-common-items/blob/main/tests/test_pages_price.py#L188

This PR fixes two of these:
* https://github.com/zytedata/zyte-common-items/blob/6620a1538ee3b558d78705800ad4ea19f058f7dc/tests/test_pages_price.py#L69
* https://github.com/zytedata/zyte-common-items/blob/6620a1538ee3b558d78705800ad4ea19f058f7dc/tests/test_pages_price.py#L71
While three are same as before:
* https://github.com/zytedata/zyte-common-items/blob/6620a1538ee3b558d78705800ad4ea19f058f7dc/tests/test_pages_price.py#L41
* https://github.com/zytedata/zyte-common-items/blob/6620a1538ee3b558d78705800ad4ea19f058f7dc/tests/test_pages_price.py#L80
* https://github.com/zytedata/zyte-common-items/blob/6620a1538ee3b558d78705800ad4ea19f058f7dc/tests/test_pages_price.py#L188

It seems like an improvement, albeit incomplete (but fixing this for all cases seems kind of out of scope so perhaps we can live with it for now...?)

File: zyte_common_items/processors.py [hunk 4]
```diff
@@ -330,6 +363,45 @@ def aggregateRating(self):
     return value
 
 
+def images_processor(value: Any, page: Any) -> Any:
```
Comment (kmike) [inline zyte_common_items/processors.py:366]: it seems page argument is unused - can we remove it?

File: zyte_common_items/processors.py [hunk 5]
```diff
@@ -330,6 +363,45 @@ def aggregateRating(self):
...
+
+    If the input is a string, it's used as a url for returning image object.
+
+    If input is either an iterable of strings or mappings with "url" key, they are
+    used to populate image objects.
+
+    Other inputs are returned unchanged.
+    """
+
+    value = _handle_selectorlist(value)
```
Comment (kmike) [inline zyte_common_items/processors.py:378]: Why is this needed? It seems that if a SelectorList is passed, the first item is taken, and then returned as-is. But it doesn't seem to be really supported, as in the end it's not a valid Image instance (as you have in TODO). So, it seems for now it can be dropped.

File: zyte_common_items/processors.py (hunk 4 above)
Comment (kmike) [reply zyte_common_items/processors.py:366]: Although I see other processors are written this way as well; it makes sense to keep it for consistency for now, and address later if needed.

File: zyte_common_items/processors.py [hunk 6]
```diff
@@ -104,50 +107,80 @@ def _from_zp_breadcrumb(value: zp_Breadcrumb) -> Breadcrumb:
...
     Other inputs are returned as is.
 
     .. _price-parser: https://github.com/scrapinghub/price-parser
     """
-    price = extract_price(value)
-    return _format_price(price)
+    value = _handle_selectorlist(value)
+
+    if isinstance(value, Real):
+        return f"{value:.2f}"
```
Comment (kmike) [inline zyte_common_items/processors.py:178]: What do you think about moving it to https://github.com/zytedata/zyte-parsers/blob/main/zyte_parsers/price.py?

Comment (kmike) [reply zyte_common_items/processors.py:178]: It seems this would allow to remove the duplication in price_processor vs simple_price_processor.

File: tests/test_processors.py [hunk 7]
```diff
@@ -149,7 +159,7 @@ def brand(self):
 
 def test_brand_page():
     class MyProductPage(ProductPage):
-        @field
+        @field(out=[brand_processor])
```
Comment (kmike) [inline tests/test_processors.py:162]: why is this needed? Shouldn't subclasses of ProductPage have brand_processor enabled by default?

File: tests/test_processors.py [hunk 8]
```diff
@@ -321,3 +331,76 @@ def aggregateRating(self):
...
+            [
+                Image("https://www.url.com/img1.jpg"),
+                Image("https://www.url.com/img2.jpg"),
+            ],
+        ),
+    ],
+)
+def test_images(input_value, expected_value):
+    class ImagesPage(BasePage):
+        @field(out=[images_processor])
```
Comment (kmike) [inline tests/test_processors.py:373]: Does it work without `out`, as images_processor is enabled by default in this PR? It'd be nice to test it without `out`, because it's going to be the most important use case (default behavior).

File: tests/test_processors.py [hunk 9]
```diff
@@ -321,3 +331,76 @@ def aggregateRating(self):
...
+        ("buy 10 ab ab", "10.00"),
+        ("1,000.17", "1000.17"),
+        ("1,000", "1000.00"),
+        (22.9, "22.90"),
+        (22.0, "22.00"),
+    ],
+)
+def test_prices(input_value, expected_value):
+    class PricePage(BasePage):
+        @field(out=[price_processor])
```
Comment (kmike) [inline tests/test_processors.py:401]: is `out` needed?

File: tests/test_pages_price.py [hunk 10]
```diff
@@ -62,21 +62,21 @@ def price(self):
     url = "https://example.com"
     page = CustomProductPage(response=HttpResponse(url=url, body=html))
     assert page.call_count == 0
-    assert page.price == "$13.2"
+    assert page.price == "13.20"
```
Comment (kmike) [inline tests/test_pages_price.py:65]: This is a big change, because it goes against the documentation - see https://zyte-common-items.readthedocs.io/en/latest/usage/field-processors.html#overview:

> By design, the processors enabled by default are “transparent”: they don’t change the output of the field if the result is of the expected final type. For example, if there is a [str](https://docs.python.org/3/library/stdtypes.html#str) attribute in the item, and the field returns [str](https://docs.python.org/3/library/stdtypes.html#str) value, the default processor returns the value as-is.

This design decision was made for two reasons:

1. It makes it possible to safely add new default processors, without a risk of breaking user code. If the code was working before a processor is introduced, it means the field was returning the final data type. Processor won't touch it, so no breaking is possible.
2. it provides an escape hatch for the user; if you want to return a final data, you can do it just by returning the value.

It's not an issue for Real support in price parsing, for Brand processing, or for Images (it all looks good in this PR), but it's an issue for price fields.

Obviously, this "design principle" has a downside: in most cases, we do want to process strings by default. I think most large Scrapy projects develop this kind of utilities over time, where all values are processed by default. We do want to make it more standard.

I can see 2 ways to address it. 

1. Start procesing final data types by default; drop this design principle.
2. Introduce more aggressive / complete versions of procesors. Keep "cautious" processors enabled by default, which don't handle the final data types. Make it easy to enable the complete versions - either by defining Processors class which users can use, or by providing another base ProductPage class, with processing of final values enabled by default.

I'm +0.5 for option (2), but willing to hear the opinions and arguments. I like keeping the "safe" approach, but it introduces more complexity (e.g. an additional base class, where users need to pick the right one). 

cc @wRAR @Gallaecio.

**Comment by kmike** [comment]
Overall, 👍, much needed changes. Having e.g. an images processor is a must.

File: zyte_common_items/processors.py (hunk 5 above)
Comment (Nykakin) [reply zyte_common_items/processors.py]: Moved it inside `TODO` comment in d2ab672afffdc9aad5a064a8345e663d4af8dd33 since indeed it isn't needed for now.

File: zyte_common_items/processors.py (hunk 6 above)
Comment (Nykakin) [reply zyte_common_items/processors.py:178]: We'd like to migrate our stuff quickly before certain deadlines, which is why I want to have this thing working for now and avoid coordinating between PR in yet another dependency repository (which is also why I placed proposed parsing of images into TODO). I think we can improve it later and migrate into more fitting place.

File: tests/test_processors.py (hunk 7 above)
Comment (Nykakin) [reply tests/test_processors.py:162]: Yes, replaced `ProductPage` with `BasePage` in d2ab672afffdc9aad5a064a8345e663d4af8dd33.

File: tests/test_processors.py (hunk 8 above)
Comment (Nykakin) [reply tests/test_processors.py:373]: These are tests for processors, not particular pages, which is why I think it makes sense to limit to `BasePage`. Copied this approach from other test in this file, such as [test for Gtin processor](https://github.com/zytedata/zyte-common-items/blob/main/tests/test_processors.py#L221-L228). After all, same processor can be applied to many different page types.

File: tests/test_processors.py (hunk 9 above)
Comment (Nykakin) [reply tests/test_processors.py:401]: In this case yes, base class is `BasePage` so there's no default processor for `price` field and I need to mark it explicitly.

File: tests/test_pages_price.py (hunk 10 above)
Comment (Nykakin) [reply tests/test_pages_price.py:65]: I'm kind of against this transparency principle, and I've even proposed another PR https://github.com/zytedata/zyte-common-items/pull/101 to ensure all strings are trimmed.

It's difficult to imagine for me why we would want to emit data with extra whitespaces and it's bothersome to always manually add some sort of default text cleaning processor into all used fields. I think this kind of processing should be applied by default at some point and would help to ensure all our output data respect a common standard.

File: tests/test_processors.py (hunk 7 above)
Comment (Nykakin) [reply tests/test_processors.py:163]: Actually it needs to be `Page`, not `BasePage`, because `self.css` is not working otherwise, updated in 30baf8a0ca371d93af1118a7829d8f4e607378a8

**Comment by Nykakin** [comment]
In response to https://github.com/zytedata/zyte-common-items/pull/99#discussion_r1725091519 and after some discussion in slack I've decided to just return strings without any change in `price_processor` instead of parsing it, thus retaining original philosophy of transparent processors.

The discussion about this approach can continue but I do not want to block image processors and brand processor because of it.

File: tests/test_processors.py (hunk 7 above)
Comment (Nykakin) [reply tests/test_processors.py]: Looks like code coverage tool is complaining, so in the end I had to remove `brand_processor` as originally suggested. Fixed in aa42e22b7db52fc5ab1a711df44f6180069d6d37

**Review by kmike — APPROVED** [review]
👍

File: zyte_common_items/processors.py (hunk 6 above)
Comment (kmike) [reply zyte_common_items/processors.py:176]: Sounds good!

File: tests/test_processors.py [hunk 11]
```diff
@@ -321,3 +331,83 @@ def aggregateRating(self):
...
+
+@pytest.mark.parametrize(
+    "input_value,expected_value",
+    [
+        (100, "100.00"),
+        (None, None),
+        ([], []),
+        ({}, {}),
+        (22.9, "22.90"),
+        (22.0, "22.00"),
```
Comment (Gallaecio) [inline tests/test_processors.py:403]: Should we add a test scenario that confirms that strings are left as is, e.g. `("22.9", "22.9")`?

File: zyte_common_items/processors.py [hunk 12]
```diff
@@ -104,50 +107,78 @@ def _from_zp_breadcrumb(value: zp_Breadcrumb) -> Breadcrumb:
...
+
+    If value is a string, uses it to create a :class:`~zyte_common_items.Brand` instance.
+
+    Other inputs are returned unchanged.
     """
-    return extract_brand_name(value, search_depth=2)
+    value = _handle_selectorlist(value)
 
+    if isinstance(value, str):
+        return Brand(name=value) if value else None
```
Comment (Gallaecio) [inline zyte_common_items/processors.py:124]: @kmike Do you think stripping would make sense here? Or if we want to keep the original spacing for non-empty strings, at least do `… if value.strip() else …`?

File: zyte_common_items/processors.py [hunk 13]
```diff
@@ -330,6 +361,44 @@ def aggregateRating(self):
...
+
+    Other inputs are returned unchanged.
+    """
+
+    # TODO: add generic-purpose extract_images utility to zyte-parsers
+    #
+    # value = _handle_selectorlist(value)
+    # if isinstance(value, (Selector, HtmlElement)):
+    #    images = extract_images(value)
+    #    return [Image(url=url) for url in images]
```
Comment (Gallaecio) [inline zyte_common_items/processors.py:381]: Could we move this comment into a repository issue instead?

**Review by Gallaecio — APPROVED** [review]

File: zyte_common_items/processors.py (hunk 12 above)
Comment (kmike) [reply zyte_common_items/processors.py:124]: I think stripping is fine here.

File: zyte_common_items/processors.py (hunk 13 above)
Comment (Nykakin) [reply zyte_common_items/processors.py]: Removed comment in 12cb8efadd59d399b41bf59de6b0d721d4d10efb created issue https://github.com/zytedata/zyte-parsers/issues/16

File: zyte_common_items/processors.py (hunk 12 above)
Comment (Nykakin) [reply zyte_common_items/processors.py]: Added `strip()` and some more test cases in 12cb8efadd59d399b41bf59de6b0d721d4d10efb

File: tests/test_processors.py (hunk 11 above)
Comment (Nykakin) [reply tests/test_processors.py:405]: Added new test cases in 12cb8efadd59d399b41bf59de6b0d721d4d10efb

**Comment by kmike** [comment]
Thanks @Nykakin!
